internal/engine/compiler/engine.go

```diff
@@ -1034,19 +1039,72 @@ func (ce *callEngine) builtinFunctionTableGrow(tables []*wasm.TableInstance) {
 	ce.pushValue(uint64(res))
 }
 
+type stackIterator struct {
+	stack   []uint64
+	fn      *function
+	base    int
+	started bool
+}
+
+func (si *stackIterator) Reset(stack []uint64, fn *function, base int) {
+	si.stack = stack
+	si.fn = fn
+	si.base = base
+	si.started = false
+}
+
+func (si *stackIterator) Clear() {
+	si.stack = nil
+	si.fn = nil
+	si.base = 0
+	si.started = false
+}
+
+func (si *stackIterator) Next() bool {
+	if !si.started {
+		si.started = true
+		return true
+	}
+
+	if si.fn == nil || si.base == 0 {
+		return false
+	}
+
+	frame := si.base + callFrameOffset(si.fn.funcType)
+	si.base = int(si.stack[frame+1] >> 3)
+	si.fn = (*function)(unsafe.Pointer(uintptr(si.stack[frame+2]))) // TODO safer
```

Comment:
Would love to know a better way to grab the `*function` pointer from the stackframe!

Reply:
unfortunately, this is the only way as we cannot have a typed stack. let's just comment on what's we are doing here.
```suggestion
	// *function lives in the third field of callFrame struct. This must be aligned with the definition of callFrame struct.
	si.fn = (*function)(unsafe.Pointer(uintptr(si.stack[frame+2])))
```